### PR TITLE
refactor(connlib): replace `webrtc-rs` with `snownet`

### LINKED ISSUE #3377: Replace `webrtc-rs` in `firezone-tunnel` with `snownet`

### PR DESCRIPTION
**Pull request by conectado**
Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

Resolves: #3377.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
@conectado The latest panic seems to be here: https://github.com/algesten/str0m/blob/ff331bd958bb91ad6d189e5aa32e550883c8b036/src/io/stun.rs#L82.

Are we reading a zero-length packet?

**Comment by thomaseizinger** [comment]
@conectado You need to drop acdbe8fba5300088e473610487d9c40eeb78a178 during rebase to avoid conflicts as that PR already merged now :)

I had just cherry-picked the commit.

**Comment by conectado** [comment]
> @conectado The latest panic seems to be here: https://github.com/algesten/str0m/blob/ff331bd958bb91ad6d189e5aa32e550883c8b036/src/io/stun.rs#L82.
> 
> Are we reading a zero-length packet?

There's nothing preventing us from reading zero-length packets but there's also no reason for that to be happening.

File: rust/connlib/tunnel/src/lib.rs [hunk 1]
```diff
@@ -214,8 +214,20 @@ where
             }
         }
 
-        let relay_socket_v4 = Socket::bind((IpAddr::from(Ipv4Addr::UNSPECIFIED), 0));
-        let relay_socket_v6 = Socket::bind((IpAddr::from(Ipv6Addr::UNSPECIFIED), 0));
+        // TODO: We need to make this dynamic and cannot fail on missing IPv4 or IPv6.
```
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: @thomaseizinger this is just to test CI right? The problem is we don't know which interface can reach the relays (and it can change over time). This could even happen in CI

File: rust/connlib/tunnel/src/lib.rs [hunk 2]
```diff
@@ -110,53 +78,282 @@ pub(crate) fn get_v6(ip: IpAddr) -> Option<Ipv6Addr> {
...
+{
+    fn new(private_key: StaticSecret) -> Self {
+        let if_watcher = IfWatcher::new().expect(
+            "Program should be able to list interfaces on the system. Check binary's permissions",
+        );
+
+        let mut connection_pool = Node::new(private_key, std::time::Instant::now());
+        let mut udp_sockets = UdpSockets::default();
+
+        for ip in if_watcher.iter() {
```
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: Can we do something better than this? This would give us 1 local candidate per ip. This can be a problem due to the combinatoric number of candidate pairs, that's why we used to limit it to 8. I'd like to at least limit this to 1 ipv4 and 1 ipv6 per iface but we don't have which iface the ip came from here.

**Comment by conectado** [comment]
> > @conectado The latest panic seems to be here: https://github.com/algesten/str0m/blob/ff331bd958bb91ad6d189e5aa32e550883c8b036/src/io/stun.rs#L82.
> > Are we reading a zero-length packet?
> 
> There's nothing preventing us from reading zero-length packets but there's also no reason for that to be happening.

This seems to be a regression associated with [8e5e5c5](https://github.com/firezone/firezone/pull/3391/commits/8e5e5c5f1fee2b08329d9db8b14d7532f7225dc5)

**Comment by conectado** [comment]
> > @conectado The latest panic seems to be here: https://github.com/algesten/str0m/blob/ff331bd958bb91ad6d189e5aa32e550883c8b036/src/io/stun.rs#L82.
> > Are we reading a zero-length packet?
> 
> There's nothing preventing us from reading zero-length packets but there's also no reason for that to be happening.

I think the problem is here's we're misusing what `len` [here](https://github.com/firezone/firezone/pull/3391/commits/8e5e5c5f1fee2b08329d9db8b14d7532f7225dc5#diff-13fba5ae9ebf7562ee01ddab8b0a60caba9b834ddeb070c4677745eaccb4c133R72) means.

To get the `len` of the message we need to use `meta.len` though, I still wonder why we get 0 messages sometimes if the socket is ready.
That's actually a call to [`recvmmsg`](https://man7.org/linux/man-pages/man2/recvmmsg.2.html) and [returns the number of messages](https://github.com/quinn-rs/quinn/blob/main/quinn-udp/src/unix.rs#L460)(it does have a fallback but it always returns 1 )

**Comment by conectado** [comment]
> > > @conectado The latest panic seems to be here: https://github.com/algesten/str0m/blob/ff331bd958bb91ad6d189e5aa32e550883c8b036/src/io/stun.rs#L82.
> > > Are we reading a zero-length packet?
> > 
> > 
> > There's nothing preventing us from reading zero-length packets but there's also no reason for that to be happening.
> 
> I think the problem is here's we're misusing what `len` [here](https://github.com/firezone/firezone/pull/3391/commits/8e5e5c5f1fee2b08329d9db8b14d7532f7225dc5#diff-13fba5ae9ebf7562ee01ddab8b0a60caba9b834ddeb070c4677745eaccb4c133R72) means.
> 
> To get the `len` of the message we need to use `meta.len` though, I still wonder why we get 0 messages sometimes if the socket is ready. That's actually a call to [`recvmmsg`](https://man7.org/linux/man-pages/man2/recvmmsg.2.html) and [returns the number of messages](https://github.com/quinn-rs/quinn/blob/main/quinn-udp/src/unix.rs#L460)(it does have a fallback but it always returns 1 )

Even with that change, using `meta.len` it seems we're reading a 0-length packet, though the message seems to have a few non-zero bytes at the begging

**Comment by conectado** [comment]
> > > > @conectado The latest panic seems to be here: https://github.com/algesten/str0m/blob/ff331bd958bb91ad6d189e5aa32e550883c8b036/src/io/stun.rs#L82.
> > > > Are we reading a zero-length packet?
> > > 
> > > 
> > > There's nothing preventing us from reading zero-length packets but there's also no reason for that to be happening.
> > 
> > 
> > I think the problem is here's we're misusing what `len` [here](https://github.com/firezone/firezone/pull/3391/commits/8e5e5c5f1fee2b08329d9db8b14d7532f7225dc5#diff-13fba5ae9ebf7562ee01ddab8b0a60caba9b834ddeb070c4677745eaccb4c133R72) means.
> > To get the `len` of the message we need to use `meta.len` though, I still wonder why we get 0 messages sometimes if the socket is ready. That's actually a call to [`recvmmsg`](https://man7.org/linux/man-pages/man2/recvmmsg.2.html) and [returns the number of messages](https://github.com/quinn-rs/quinn/blob/main/quinn-udp/src/unix.rs#L460)(it does have a fallback but it always returns 1 )
> 
> Even with that change, using `meta.len` it seems we're reading a 0-length packet, though the message seems to have a few non-zero bytes at the begging

Well, it seems like I'm [misinterpreting `meta.len`](https://github.com/quinn-rs/quinn/blob/main/quinn-udp/tests/tests.rs#L150-L153)

**Comment by thomaseizinger** [comment]
> > > > > @conectado The latest panic seems to be here: https://github.com/algesten/str0m/blob/ff331bd958bb91ad6d189e5aa32e550883c8b036/src/io/stun.rs#L82.
> > > > > Are we reading a zero-length packet?
> > > > 
> > > > 
> > > > There's nothing preventing us from reading zero-length packets but there's also no reason for that to be happening.
> > > 
> > > 
> > > I think the problem is here's we're misusing what `len` [here](https://github.com/firezone/firezone/pull/3391/commits/8e5e5c5f1fee2b08329d9db8b14d7532f7225dc5#diff-13fba5ae9ebf7562ee01ddab8b0a60caba9b834ddeb070c4677745eaccb4c133R72) means.
> > > To get the `len` of the message we need to use `meta.len` though, I still wonder why we get 0 messages sometimes if the socket is ready. That's actually a call to [`recvmmsg`](https://man7.org/linux/man-pages/man2/recvmmsg.2.html) and [returns the number of messages](https://github.com/quinn-rs/quinn/blob/main/quinn-udp/src/unix.rs#L460)(it does have a fallback but it always returns 1 )
> > 
> > 
> > Even with that change, using `meta.len` it seems we're reading a 0-length packet, though the message seems to have a few non-zero bytes at the begging
> 
> Well, it seems like I'm [misinterpreting `meta.len`](https://github.com/quinn-rs/quinn/blob/main/quinn-udp/tests/tests.rs#L150-L153)

What on earth is going on there 😂

File: rust/connlib/tunnel/src/lib.rs (hunk 2 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/lib.rs]: I'll do some research on this today.

File: rust/connlib/snownet/src/node.rs [hunk 3]
```diff
@@ -164,14 +166,14 @@ where
         packet: &[u8],
         now: Instant,
         buffer: &'s mut [u8],
-    ) -> Result<Option<(TId, IpPacket<'s>)>, Error> {
+    ) -> Result<Option<(TId, MutableIpPacket<'s>)>, Error> {
         // First, check if a `StunBinding` wants the packet
```
Comment (thomaseizinger) [inline rust/connlib/snownet/src/node.rs]: @conectado I am wondering, should we automatically add `local` as a host-candidate here? If we are receiving traffic on this socket, then this is most likely one of the best sockets to use.

We could then even remove `add_local_interface`, assuming we always have at least one STUN and one TURN server to talk to AND we aren't trying to connect to a gateway that is reachable over the local network but without connectivity to a STUN & TURN server.

I think that could even remove our dependency on `if-watcher`!

Comment (thomaseizinger) [reply rust/connlib/snownet/src/node.rs]: I convinced myself that this is an exceptional idea and implemented it: https://github.com/firezone/firezone/pull/3528

File: swift/apple/Firezone.xcodeproj/project.pbxproj
Comment (thomaseizinger) [inline swift/apple/Firezone.xcodeproj/project.pbxproj:1]: I think this can now be reverted again because we no longer depend on `if-watch`.

File: rust/connlib/tunnel/src/ip_packet.rs [hunk 4]
```diff
@@ -199,6 +199,16 @@ pub enum IpPacket<'a> {
     Ipv6Packet(Ipv6Packet<'a>),
 }
 
+// TODO: simply get rid of this implementation of IpPacket
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/ip_packet.rs]: Do we want to do this as part of this PR?

File: rust/connlib/tunnel/src/lib.rs [hunk 5]
```diff
@@ -110,53 +77,201 @@ pub(crate) fn get_v6(ip: IpAddr) -> Option<Ipv6Addr> {
...
+                if let Some((_, info)) = self.connections.node.stats().find(|(i, _)| id == i) {
+                    if info.missed_keep_alives() >= 2 {
+                        tracing::debug!(%id, "connection expired");
+                        Some(*id)
+                    } else {
+                        None
+                    }
+                } else {
+                    tracing::debug!(%id, "connection expired");
+                    Some(*id)
```
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: It's good to leave this here for hardening but I notice that sometimes `ConnectionFailed` event is not emitted

File: rust/connlib/tunnel/src/lib.rs [hunk 6]
```diff
@@ -110,53 +77,201 @@ pub(crate) fn get_v6(ip: IpAddr) -> Option<Ipv6Addr> {
...
+                    connection,
+                    candidate,
+                }) => {
+                    return Poll::Ready(Event::SignalIceCandidate {
+                        conn_id: connection,
+                        candidate,
+                    });
+                }
+                Some(snownet::Event::ConnectionEstablished(id)) => {
+                    tracing::info!(%id, "Connection established with peer");
```
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: This seems to be emitted thrice per connection

File: rust/gateway/src/eventloop.rs [hunk 7]
```diff
@@ -142,6 +142,8 @@ impl Eventloop {
                         (req.client.id, req.reference.clone()),
                         async move {
                             let conn = tunnel
+                                // TODO: the fact that this is async can make us
+                                // generate candidates before the connection response is generated
```
Comment (thomaseizinger) [inline rust/gateway/src/eventloop.rs]: This is due to DNS resolution in `set_peer_connection_request` right? Can we just push that up to the eventloop and delay the call to `set_peer_connection_request` until the resolution is complete?

Comment (conectado) [reply rust/gateway/src/eventloop.rs]: That's the case.

We can do that now.

But I was planning rather to send the response here and have the dns resolution be queued up and sent at a later time but I was planning to leave that for a later PR.

Comment (thomaseizinger) [reply rust/gateway/src/eventloop.rs]: > But I was planning rather to send the response here and have the dns resolution be queued up and sent at a later time but I was planning to leave that for a later PR.

If that is an option, yeah. I figured the result of the DNS response is required.

To be clear, with push up to the eventloop I meant do the DNS resolution right here where this comment is and then call `set_peer_connection_request` with the result.

Comment (thomaseizinger) [reply rust/gateway/src/eventloop.rs]: I guess it could be unified with the `allow_access` API?

Comment (conectado) [reply rust/gateway/src/eventloop.rs]: > I guess it could be unified with the allow_access API?

Yeah, precisely, I think it's a good idea long-term to separate the gateway connection from the resource authorization

File: rust/connlib/snownet/src/allocation.rs [hunk 8]
```diff
@@ -766,7 +767,7 @@ impl ChannelBindings {
         let channel = self.inner.get_mut(&channel_number)?;
         channel.record_received(now);
 
-        debug_assert!(channel.bound); // TODO: Should we "force-set" this? We seem to be getting traffic on this channel ..
+        // debug_assert!(channel.bound); // TODO: Should we "force-set" this? We seem to be getting traffic on this channel ..
```
Comment (conectado) [inline rust/connlib/snownet/src/allocation.rs]: @thomaseizinger this was failing when I was testing but everything worked anyways when commenting it out

File: rust/connlib/tunnel/src/control_protocol/client.rs [hunk 9]
```diff
@@ -201,12 +206,39 @@ where
...
+                    }
+                    itertools::EitherOrBoth::Left(external_ip) => peer
+                        .transform
+                        .get_or_assign_translation(external_ip, &mut role_state.ip_provider),
+                    // We arbitrarily duplicate the first ip.
+                    // TODO: think about other strategies to handle this.
+                    itertools::EitherOrBoth::Right(proxy_ip) => {
+                        peer.transform.assign_translation(&first_ip, proxy_ip)
+                    }
+                }
```
Comment (conectado) [inline rust/connlib/tunnel/src/control_protocol/client.rs]: When this is finished this fixes #3369

File: rust/gateway/src/eventloop.rs (hunk 7 above)
Comment (thomaseizinger) [reply rust/gateway/src/eventloop.rs]: I'll try and do that on `main` time-boxed because it does help with the integration here on getting rid of one more `async` :)

File: rust/connlib/snownet/src/allocation.rs (hunk 8 above)
Comment (thomaseizinger) [reply rust/connlib/snownet/src/allocation.rs]: I'll look into this. From memory it is a race condition of our local state that I thought could never happen.

File: rust/gateway/src/eventloop.rs (hunk 7 above)
Comment (thomaseizinger) [reply rust/gateway/src/eventloop.rs]: Done, didn't take long: https://github.com/firezone/firezone/pull/3622

File: rust/connlib/snownet/src/allocation.rs (hunk 8 above)
Comment (thomaseizinger) [reply rust/connlib/snownet/src/allocation.rs]: This is interesting. If this `debug_assert` fails then it means we are receiving traffic on this channel but the relay has not yet given us a success response for the binding. Can you share logs from when this got triggered? It seems very unlikely that the relay would not send us a success response but relay data to us.

File: rust/connlib/tunnel/src/device_channel/tun_linux.rs
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/device_channel/tun_linux.rs:1]: @conectado Should we move the changes in here to `main` before merging this? I think that is unrelated, right?

Comment (conectado) [reply rust/connlib/tunnel/src/device_channel/tun_linux.rs:1]: not completely unrelated, the routing loop this prevents only happens now because we now create the peer before the connection is established.

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/device_channel/tun_linux.rs:1]: Ah okay, I didn't make that connection :)

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/device_channel/tun_linux.rs:1]: Couldn't we fix this by reacting to the `ConnectionEstablished` event in the gateway?

File: elixir/apps/api/lib/api/client/views/relay.ex
Comment (thomaseizinger) [inline elixir/apps/api/lib/api/client/views/relay.ex:1]: @conectado Even if it is not much, I think I'd prefer to make this change to the portal in a separate PR later instead of with this PR.

File: rust/connlib/tunnel/src/device_channel/tun_linux.rs
Comment (conectado) [reply rust/connlib/tunnel/src/device_channel/tun_linux.rs:1]: I think preventing routing loops is still the way to go, they can happen in case of re-connections and it's less state to track

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/device_channel/tun_linux.rs:1]: Hmm okay but that sounds like it would be useful in isolation too, i.e. could be extracted into its own PR?

File: rust/connlib/tunnel/src/sockets.rs [hunk 10]
```diff
@@ -191,20 +192,49 @@ impl<const N: usize> Socket<N> {
...
+
+#[cfg(target_os = "android")]
+fn protect(
+    socket: &socket2::Socket,
+    callbacks: &CallbackErrorFacade<impl Callbacks>,
+) -> Result<()> {
+    use connlib_shared::Callbacks;
+    use std::os::fd::AsRawFd;
+
+    callbacks.protect_file_descriptor(socket.as_raw_fd())?;
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/sockets.rs]: Is this strictly necessary? We didn't do that before and I have some reservations about introducing more callbacks :)

See https://github.com/firezone/firezone/issues/3458.

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/sockets.rs]: If it is necessary, I think it would be cleaner to (for now) instead have getters on `Tunnel` to expose the file descriptors in order for the android client code to call `protect` on it.

At a later point, we can then explore doing something like what I outlined in #3458. At least, that would avoid passing the callbacks down to low-level code like this.

Another alternative would be to make the `Sockets` an argument of the tunnel. That way, the android glue code could configure those before constructing the tunnel.

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/sockets.rs]: Let me know what you think of [5973943](https://github.com/firezone/firezone/pull/3391/commits/5973943c3d1ddb62a1b0b2a8f478d304cfc917fe).

**Review by thomaseizinger — APPROVED** [review]
